Index: src/game/g_admin.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_admin.c,v
retrieving revision 1.1.1.1.4.19
diff -u -p -r1.1.1.1.4.19 g_admin.c
--- src/game/g_admin.c	17 Feb 2009 17:10:43 -0000	1.1.1.1.4.19
+++ src/game/g_admin.c	17 Feb 2009 17:21:09 -0000
@@ -2495,6 +2495,14 @@ qboolean G_admin_kick( gentity_t *ent, i
   va( "disconnect \"You have been kicked.\n%s^7\nreason:\n%s\n%s\"",
     ( ent ) ? va( "admin:\n%s", G_admin_adminPrintName( ent ) ) : "admin\nconsole",
     ( *reason ) ? reason : "kicked by admin", notice ) );
+
+  G_LogPrintf( "kick: %i %i [%s] (%s) %s^7 %s^7\n",
+    vic->client->ps.clientNum,
+    G_admin_parse_time( g_adminTempBan.string ),
+    vic->client->pers.ip,
+    vic->client->pers.guid,
+    vic->client->pers.netname,
+    ( *reason ) ? reason : "automatic temp ban created by kick" );
   
   trap_DropClient( pids[ 0 ], va( "kicked%s^7, reason: %s",
     ( ent ) ? va( " by %s", G_admin_adminPrintName( ent ) ) : " by console",
@@ -2686,6 +2694,13 @@ qboolean G_admin_ban( gentity_t *ent, in
   else
     admin_writeconfig();
 
+  G_LogPrintf( "ban: %i %i [%s] (%s) %s^7 %s^7\n",
+    g_admin_namelog[ logmatch ]->slot, seconds,
+    g_admin_namelog[ logmatch ]->ip,
+    g_admin_namelog[ logmatch ]->guid,
+    g_admin_namelog[ logmatch ]->name[ 0 ],
+    ( *reason ) ? reason : "banned by admin" );
+
   if( g_admin_namelog[ logmatch ]->slot == -1 ) 
   {
     // client is already disconnected so stop here
