Index: src/game/g_admin.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_admin.c,v
retrieving revision 1.1.1.1.4.11
diff -u -p -r1.1.1.1.4.11 g_admin.c
--- src/game/g_admin.c	17 Feb 2009 13:38:02 -0000	1.1.1.1.4.11
+++ src/game/g_admin.c	17 Feb 2009 13:42:05 -0000
@@ -280,8 +280,8 @@ g_admin_cmd_t g_admin_cmds[ ] = 
      //kev: a bit of a hack, but there is no real point to
      //creating a new admin flag for this, so i stole it from !help
     {"specme", G_admin_putmespec, "h",
-        "moves you to the spectators",
-    ""
+      "moves you to the spectators (can be done silently with the 's' argument)",
+      "(^5s^7)"
     },
 
     {"subnetban", G_admin_subnetban, "E",
@@ -5855,6 +5855,20 @@ qboolean G_admin_putmespec( gentity_t *e
   }
   
   G_ChangeTeam( ent, PTE_NONE );
+
+  // check for silent '!specme s' - requires !kick permission
+  if( G_SayArgc() > 1 + skiparg )
+  {
+    char arg[ 2 ];
+
+    G_SayArgv( 1 + skiparg, arg, sizeof( arg ) );
+    if( ( arg[ 0 ] == 's' || arg[ 0 ] == 'S' ) && G_admin_permission( ent, 'k' ) )
+    {
+      ADMP("^3!specme: ^7 You have silently joined the spectators\n");
+      return qtrue;
+    }
+  }
+
   AP( va("print \"^3!specme: ^7%s^7 decided to join the spectators\n\"", ent->client->pers.netname ) );
   return qtrue;
 }
