Index: src/game/g_cmds.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_cmds.c,v
retrieving revision 1.1.1.1.4.17
diff -u -p -r1.1.1.1.4.17 g_cmds.c
--- src/game/g_cmds.c	17 Feb 2009 19:51:17 -0000	1.1.1.1.4.17
+++ src/game/g_cmds.c	17 Feb 2009 20:08:19 -0000
@@ -4250,6 +4250,119 @@ char *G_statsString( statsCounters_t *sc
 }
 
 
+/*
+=================
+Cmd_TeamStatus_f
+=================
+*/
+void Cmd_TeamStatus_f( gentity_t *ent )
+{
+  char multiple[ 12 ];
+  int builders = 0;
+  int arm = 0, mediboost = 0;
+  int omrccount = 0, omrchealth = 0;
+  qboolean omrcbuild = qfalse;
+  gentity_t *tmp;
+  int i;
+
+  if( !g_teamStatus.integer )
+  {
+    trap_SendServerCommand( ent - g_entities,
+      "print \"teamstatus is disabled.\n\"" );
+    return;
+  }
+
+  if( ent->client->pers.muted )
+  {
+    trap_SendServerCommand( ent - g_entities,
+      "print \"You are muted and cannot use message commands.\n\"" );
+    return;
+  }
+
+  if( ent->client->pers.lastTeamStatus && 
+      ( level.time - ent->client->pers.lastTeamStatus) < g_teamStatus.integer * 1000 )
+  {
+    ADMP( va("You may only check your team's status once every %i seconds.\n",
+          g_teamStatus.integer  ));
+    return;
+  }
+
+  ent->client->pers.lastTeamStatus = level.time;
+
+  tmp = &g_entities[ 0 ];
+  for ( i = 0; i < level.num_entities; i++, tmp++ )
+  {
+    if( i < MAX_CLIENTS )
+    {
+      if( tmp->client &&
+          tmp->client->pers.connected == CON_CONNECTED &&
+          tmp->client->pers.teamSelection == ent->client->pers.teamSelection &&
+          tmp->health > 0 &&
+          ( tmp->client->ps.stats[ STAT_PCLASS ] == PCL_ALIEN_BUILDER0 ||
+            tmp->client->ps.stats[ STAT_PCLASS ] == PCL_ALIEN_BUILDER0_UPG ||
+            BG_InventoryContainsWeapon( WP_HBUILD, tmp->client->ps.stats ) ||
+            BG_InventoryContainsWeapon( WP_HBUILD2, tmp->client->ps.stats ) ) )
+        builders++;
+      continue;
+    }
+
+    if( tmp->s.eType == ET_BUILDABLE )
+    {
+      if( tmp->biteam != ent->client->pers.teamSelection ||
+          tmp->health <= 0 )
+        continue;
+
+      switch( tmp->s.modelindex )
+      {
+        case BA_H_REACTOR:
+        case BA_A_OVERMIND:
+          omrccount++;
+          if( tmp->health > omrchealth )
+            omrchealth = tmp->health;
+          if( !omrcbuild )
+            omrcbuild = tmp->spawned;
+          break;
+        case BA_H_ARMOURY:
+          arm++;
+          break;
+        case BA_H_MEDISTAT:
+        case BA_A_BOOSTER:
+          mediboost++;
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  if( omrccount > 1 )
+    Com_sprintf( multiple, sizeof( multiple ), "^7[x%d]", omrccount );
+  else
+    multiple[ 0 ] = '\0';
+
+  if( ent->client->pers.teamSelection == PTE_ALIENS )
+  {
+    G_Say( ent, NULL, SAY_TEAM,
+      va( "^3OM: %s(%d)%s ^3Spawns: ^5%d ^3Builders: ^5%d ^3Boosters: ^5%d^7" ,
+      ( !omrccount ) ? "^1Down" : ( omrcbuild ) ? "^2Up" : "^5Building",
+      omrchealth * 100 / OVERMIND_HEALTH,
+      multiple,
+      level.numAlienSpawns,
+      builders,
+      mediboost ) );
+  }
+  else
+  {
+    G_Say( ent, NULL, SAY_TEAM,
+      va( "^3RC: %s(%d)%s ^3Spawns: ^5%d ^3Builders: ^5%d ^3Armouries: ^5%d ^3Medistations: ^5%d^7" ,
+      ( !omrccount ) ? "^1Down" : ( omrcbuild ) ? "^2Up" : "^5Building",
+      omrchealth * 100 / REACTOR_HEALTH,
+      multiple,
+      level.numHumanSpawns,
+      builders,
+      arm, mediboost ) );
+  }
+}
 
 /*
 =================
@@ -5022,6 +5135,7 @@ commands_t cmds[ ] = {
 
   { "score", CMD_INTERMISSION, ScoreboardMessage },
   { "mystats", CMD_TEAM|CMD_INTERMISSION, Cmd_MyStats_f },
+  { "teamstatus", CMD_TEAM, Cmd_TeamStatus_f },
 
   // cheats
   { "give", CMD_CHEAT|CMD_TEAM|CMD_LIVING, Cmd_Give_f },
Index: src/game/g_local.h
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_local.h,v
retrieving revision 1.1.1.1.4.24
diff -u -p -r1.1.1.1.4.24 g_local.h
--- src/game/g_local.h	17 Feb 2009 20:01:46 -0000	1.1.1.1.4.24
+++ src/game/g_local.h	17 Feb 2009 20:08:19 -0000
@@ -404,6 +404,8 @@ typedef struct
   int                 credit;
   int                 ping;
 
+  int                 lastTeamStatus;
+
   int                 lastFloodTime;         // level.time of last flood-limited command
   int                 floodDemerits;         // number of flood demerits accumulated
 
@@ -1445,6 +1447,7 @@ extern  vmCvar_t  g_privateMessages;
 extern  vmCvar_t  g_decolourLogfiles;
 extern  vmCvar_t  g_publicSayadmins;
 extern  vmCvar_t  g_myStats;
+extern  vmCvar_t  g_teamStatus;
 extern  vmCvar_t  g_antiSpawnBlock;
 
 extern  vmCvar_t  g_killingSpree;
Index: src/game/g_main.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_main.c,v
retrieving revision 1.1.1.1.4.26
diff -u -p -r1.1.1.1.4.26 g_main.c
--- src/game/g_main.c	17 Feb 2009 20:01:46 -0000	1.1.1.1.4.26
+++ src/game/g_main.c	17 Feb 2009 20:08:19 -0000
@@ -187,6 +187,7 @@ vmCvar_t  g_decolourLogfiles;
 vmCvar_t  g_minLevelToSpecMM1;
 vmCvar_t  g_publicSayadmins;
 vmCvar_t  g_myStats;
+vmCvar_t  g_teamStatus;
 vmCvar_t  g_antiSpawnBlock;
 
 vmCvar_t  g_killingSpree;
@@ -435,6 +436,7 @@ static cvarTable_t   gameCvarTable[ ] =
   { &g_decolourLogfiles, "g_decolourLogfiles", "0", CVAR_ARCHIVE, 0, qfalse  },
   { &g_buildLogMaxLength, "g_buildLogMaxLength", "50", CVAR_ARCHIVE, 0, qfalse  },
   { &g_myStats, "g_myStats", "1", CVAR_ARCHIVE, 0, qfalse  },
+  { &g_teamStatus, "g_teamStatus", "0", CVAR_ARCHIVE, 0, qfalse  },
   { &g_publicSayadmins, "g_publicSayadmins", "1", CVAR_ARCHIVE, 0, qfalse  },    
   { &g_minLevelToSpecMM1, "g_minLevelToSpecMM1", "0", CVAR_ARCHIVE, 0, qfalse  },
   { &g_antiSpawnBlock, "g_antiSpawnBlock", "0", CVAR_ARCHIVE, 0, qfalse  },
