Index: src/game/g_admin.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_admin.c,v
retrieving revision 1.1.1.1.4.5
diff -u -p -r1.1.1.1.4.5 g_admin.c
--- src/game/g_admin.c	17 Feb 2009 04:14:04 -0000	1.1.1.1.4.5
+++ src/game/g_admin.c	17 Feb 2009 04:29:13 -0000
@@ -288,6 +288,11 @@ g_admin_cmd_t g_admin_cmds[ ] = 
       "show the current local server time",
       ""},
 
+    {"tklog", G_admin_tklog, "t",
+      "list recent teamkill activity",
+      "(^5start id#|name|-skip#^7) (^5search skip#^7)"
+    },
+
     {"unban", G_admin_unban, "b",
       "unbans a player specified by the slot as seen in showbans",
       "[^3ban#^7]"
@@ -332,6 +337,9 @@ g_admin_namelog_t *g_admin_namelog[ MAX_
 static int admin_adminlog_index = 0;
 g_admin_adminlog_t *g_admin_adminlog[ MAX_ADMIN_ADMINLOGS ];
 
+static int admin_tklog_index = 0;
+g_admin_tklog_t *g_admin_tklog[ MAX_ADMIN_TKLOGS ];
+
 // This function should only be used directly when the client is connecting and thus has no GUID.
 // Else, use G_admin_permission() 
 qboolean G_admin_permission_guid( char *guid, char flag )
@@ -2973,6 +2981,250 @@ qboolean G_admin_adminlog( gentity_t *en
   return qtrue;
 }
 
+void G_admin_tklog_cleanup( void )
+{
+  int i;
+
+  for( i = 0; i < MAX_ADMIN_TKLOGS && g_admin_tklog[ i ]; i++ )
+  {
+    G_Free( g_admin_tklog[ i ] );
+    g_admin_tklog[ i ] = NULL;
+  }
+
+  admin_tklog_index = 0;
+}
+
+void G_admin_tklog_log( gentity_t *attacker, gentity_t *victim, int meansOfDeath )
+{
+  g_admin_tklog_t *tklog;
+  int previous;
+  int count = 1;
+
+  if( !attacker || !victim )
+    return;
+
+  previous = admin_tklog_index - 1;
+  if( previous < 0 )
+    previous = MAX_ADMIN_TKLOGS - 1;
+
+  if( g_admin_tklog[ previous ] )
+    count = g_admin_tklog[ previous ]->id + 1;
+
+  if( g_admin_tklog[ admin_tklog_index ] )
+    tklog = g_admin_tklog[ admin_tklog_index ];
+  else
+    tklog = G_Alloc( sizeof( g_admin_tklog_t ) );
+
+  memset( tklog, 0, sizeof( g_admin_tklog_t ) );
+  tklog->id = count;
+  tklog->time = level.time - level.startTime;
+  Q_strncpyz( tklog->name, attacker->client->pers.netname, sizeof( tklog->name ) );
+  Q_strncpyz( tklog->victim, victim->client->pers.netname, sizeof( tklog->victim ) );
+  tklog->damage = victim->client->tkcredits[ attacker->s.number ];
+  tklog->value = victim->client->ps.stats[ STAT_MAX_HEALTH ];
+  tklog->team = attacker->client->ps.stats[ STAT_PTEAM ];
+  if( meansOfDeath == MOD_GRENADE )
+    tklog->weapon = WP_GRENADE;
+  else if( tklog->team == PTE_HUMANS )
+    tklog->weapon = attacker->s.weapon;
+  else
+    tklog->weapon = attacker->client->ps.stats[ STAT_PCLASS ];
+
+  g_admin_tklog[ admin_tklog_index ] = tklog;
+  admin_tklog_index++;
+  if( admin_tklog_index >= MAX_ADMIN_TKLOGS )
+    admin_tklog_index = 0;
+}
+
+qboolean G_admin_tklog( gentity_t *ent, int skiparg )
+{
+  g_admin_tklog_t *results[ 10 ];
+  int result_index = 0;
+  char *search_name = NULL;
+  int index;
+  int skip = 0;
+  int skipped = 0;
+  int checked = 0;
+  char n1[ MAX_NAME_LENGTH ];
+  char fmt_name[ 16 ];
+  char argbuf[ 32 ];
+  char *weaponName;
+  int name_length = 10;
+  int max_id = 0;
+  int i;
+  qboolean match;
+
+  memset( results, 0, sizeof( results ) );
+
+  index = admin_tklog_index;
+  for( i = 0; i < 10; i++ )
+  {
+    int prev;
+
+    prev = index - 1;
+    if( prev < 0 )
+      prev = MAX_ADMIN_TKLOGS - 1;
+    if( !g_admin_tklog[ prev ] )
+      break;
+    if( g_admin_tklog[ prev ]->id > max_id )
+      max_id = g_admin_tklog[ prev ]->id;
+    index = prev;
+  }
+
+  if( G_SayArgc() > 1 + skiparg )
+  {
+    G_SayArgv( 1 + skiparg, argbuf, sizeof( argbuf ) );
+    if( ( *argbuf >= '0' && *argbuf <= '9' ) || *argbuf == '-' )
+    {
+      int id;
+
+      id = atoi( argbuf );
+      if( id < 0 )
+        id += ( max_id - 9 );
+      else if( id <= max_id - MAX_ADMIN_TKLOGS )
+        id = max_id - MAX_ADMIN_TKLOGS + 1;
+
+      if( id + 9 >= max_id )
+        id = max_id - 9;
+      if( id < 1 )
+        id = 1;
+      for( i = 0; i < MAX_ADMIN_TKLOGS; i++ )
+      {
+        if( g_admin_tklog[ i ]->id == id )
+        {
+          index = i;
+          break;
+        }
+      }
+    }
+    else
+    {
+      search_name = argbuf;
+    }
+
+    if( G_SayArgc() > 2 + skiparg && ( search_name ) )
+    {
+      char skipbuf[ 4 ];
+      G_SayArgv( 2 + skiparg, skipbuf, sizeof( skipbuf ) );
+      skip = atoi( skipbuf );
+    }
+  }
+
+  if( search_name )
+  {
+    g_admin_tklog_t *result_swap[ 10 ];
+
+    memset( &result_swap, 0, sizeof( result_swap ) );
+
+    index = admin_tklog_index - 1;
+    if( index < 0 )
+      index = MAX_ADMIN_TKLOGS - 1;
+
+    while( g_admin_tklog[ index ] &&
+      checked < MAX_ADMIN_TKLOGS &&
+      result_index < 10 )
+    {
+      match = qfalse;
+
+      G_SanitiseString( g_admin_tklog[ index ]->name, n1, sizeof( n1 ) );
+      if( strstr( n1, search_name ) )
+        match = qtrue;
+
+      if( match && skip > 0 )
+      {
+        match = qfalse;
+        skip--;
+        skipped++;
+      }
+      if( match )
+      {
+        result_swap[ result_index ] = g_admin_tklog[ index ];
+        result_index++;
+      }
+
+      checked++;
+      index--;
+      if( index < 0 )
+        index = MAX_ADMIN_TKLOGS - 1;
+    }
+    // search runs backwards, turn it around
+    for( i = 0; i < result_index; i++ )
+      results[ i ] = result_swap[ result_index - i - 1 ];
+  }
+  else
+  {
+    while( g_admin_tklog[ index ] && result_index < 10 )
+    {
+      results[ result_index ] = g_admin_tklog[ index ];
+      result_index++;
+      index++;
+      if( index >= MAX_ADMIN_TKLOGS )
+        index = 0;
+    }
+  }
+
+  for( i = 0; results[ i ] && i < 10; i++ )
+  {
+    int l;
+
+    G_DecolorString( results[ i ]->name, n1 );
+    l = strlen( n1 );
+    if( l > name_length )
+      name_length = l;
+  }
+  ADMBP_begin( );
+  for( i = 0; results[ i ] && i < 10; i++ )
+  {
+    int t;
+
+    t = results[ i ]->time / 1000;
+
+    G_DecolorString( results[ i ]->name, n1 );
+    Com_sprintf( fmt_name, sizeof( fmt_name ), "%%%ds", 
+      ( name_length + strlen( results[ i ]->name ) - strlen( n1 ) ) );
+    Com_sprintf( n1, sizeof( n1 ), fmt_name, results[ i ]->name );
+
+    if( results[ i ]->team == PTE_HUMANS )
+      weaponName = BG_FindNameForWeapon( results[ i ]->weapon );
+    else
+      weaponName = BG_FindNameForClassNum( results[ i ]->weapon );
+
+    ADMBP( va( "^7%3d %3d:%02d %s^7 %3d / %3d %10s %s^7\n",
+      results[ i ]->id,
+      t / 60, t % 60,
+      n1,
+      results[ i ]->damage,
+      results[ i ]->value,
+      weaponName,
+      results[ i ]->victim ) );
+  }
+  if( search_name )
+  {
+    ADMBP( va( "^3!tklog:^7 Showing %d matches for '%s^7'.",
+      result_index,
+      argbuf ) );
+    if( checked < MAX_ADMIN_TKLOGS && g_admin_tklog[ checked ] )
+      ADMBP( va( " run '!tklog %s^7 %d' to see more",
+       argbuf,
+       skipped + result_index ) );
+    ADMBP( "\n" );
+  }
+  else if ( results[ 0 ] )
+  {
+    ADMBP( va( "^3!tklog:^7 Showing %d - %d of %d.\n",
+      results[ 0 ]->id,
+      results[ result_index - 1 ]->id,
+      max_id ) );
+  }
+  else
+  {
+    ADMBP( "^3!tklog:^7 log is empty.\n" );
+  }
+  ADMBP_end( );
+
+  return qtrue;
+}
+
 qboolean G_admin_map( gentity_t *ent, int skiparg )
 {
   char map[ MAX_QPATH ];
Index: src/game/g_admin.h
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_admin.h,v
retrieving revision 1.1.1.1.4.3
diff -u -p -r1.1.1.1.4.3 g_admin.h
--- src/game/g_admin.h	17 Feb 2009 04:14:04 -0000	1.1.1.1.4.3
+++ src/game/g_admin.h	17 Feb 2009 04:27:04 -0000
@@ -38,6 +38,7 @@ Foundation, Inc., 51 Franklin St, Fifth 
 #define MAX_ADMIN_NAMELOG_NAMES 5
 #define MAX_ADMIN_ADMINLOGS 128
 #define MAX_ADMIN_ADMINLOG_ARGS 50
+#define MAX_ADMIN_TKLOGS 64
 #define MAX_ADMIN_FLAGS 64
 #define MAX_ADMIN_COMMANDS 64
 #define MAX_ADMIN_CMD_LEN 20
@@ -162,6 +163,19 @@ typedef struct g_admin_adminlog
 }
 g_admin_adminlog_t;
 
+typedef struct g_admin_tklog
+{
+  char      name[ MAX_NAME_LENGTH ];
+  char      victim[ MAX_NAME_LENGTH ];
+  int       id;
+  int       time;
+  int       damage;
+  int       value;
+  int       team;
+  int       weapon;
+}
+g_admin_tklog_t;
+
 qboolean G_admin_ban_check( char *userinfo, char *reason, int rlen );
 qboolean G_admin_cmd_check( gentity_t *ent, qboolean say );
 qboolean G_admin_readconfig( gentity_t *ent, int skiparg );
@@ -186,6 +200,9 @@ qboolean G_admin_putteam( gentity_t *ent
 qboolean G_admin_adminlog( gentity_t *ent, int skiparg );
 void G_admin_adminlog_cleanup( void );
 void G_admin_adminlog_log( gentity_t *ent, char *command, char *args, int skiparg, qboolean success );
+qboolean G_admin_tklog( gentity_t *ent, int skiparg );
+void G_admin_tklog_cleanup( void );
+void G_admin_tklog_log( gentity_t *attacker, gentity_t *victim, int meansOfDeath );
 qboolean G_admin_listadmins( gentity_t *ent, int skiparg );
 qboolean G_admin_listlayouts( gentity_t *ent, int skiparg );
 qboolean G_admin_listplayers( gentity_t *ent, int skiparg );
Index: src/game/g_combat.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_combat.c,v
retrieving revision 1.1.1.1.4.2
diff -u -p -r1.1.1.1.4.2 g_combat.c
--- src/game/g_combat.c	17 Feb 2009 04:25:09 -0000	1.1.1.1.4.2
+++ src/game/g_combat.c	17 Feb 2009 04:27:04 -0000
@@ -224,6 +224,7 @@ void player_die( gentity_t *self, gentit
       va( "cp \"You killed ^1TEAMMATE^7 %s\"", self->client->pers.netname ) );
     G_LogOnlyPrintf("%s^7 was killed by ^1TEAMMATE^7 %s^7 (Did %d damage to %d max)\n",
       self->client->pers.netname, attacker->client->pers.netname, self->client->tkcredits[ attacker->s.number ], self->client->ps.stats[ STAT_MAX_HEALTH ] );
+    G_admin_tklog_log( attacker, self, meansOfDeath );
   }
 
   self->enemy = attacker;
Index: src/game/g_main.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_main.c,v
retrieving revision 1.1.1.1.4.5
diff -u -p -r1.1.1.1.4.5 g_main.c
--- src/game/g_main.c	17 Feb 2009 04:14:04 -0000	1.1.1.1.4.5
+++ src/game/g_main.c	17 Feb 2009 04:27:04 -0000
@@ -813,6 +813,7 @@ void G_ShutdownGame( int restart )
   G_admin_cleanup( );
   G_admin_namelog_cleanup( );
   G_admin_adminlog_cleanup( );
+  G_admin_tklog_cleanup( );
 
   level.restarted = qfalse;
   level.surrenderTeam = PTE_NONE;
