Index: src/game/g_local.h
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_local.h,v
retrieving revision 1.1.1.1.4.1
diff -u -p -r1.1.1.1.4.1 g_local.h
--- src/game/g_local.h	17 Feb 2009 02:38:26 -0000	1.1.1.1.4.1
+++ src/game/g_local.h	17 Feb 2009 03:11:02 -0000
@@ -1360,6 +1360,8 @@ extern  vmCvar_t  g_mapvoteMaxTime;
 
 extern  vmCvar_t  g_msg;
 extern  vmCvar_t  g_msgTime;
+extern  vmCvar_t  g_welcomeMsg;
+extern  vmCvar_t  g_welcomeMsgTime;
 
 extern  vmCvar_t  g_buildLogMaxLength;
 
Index: src/game/g_main.c
===================================================================
RCS file: /devel/cvs/tremulous/src/game/g_main.c,v
retrieving revision 1.1.1.1.4.2
diff -u -p -r1.1.1.1.4.2 g_main.c
--- src/game/g_main.c	17 Feb 2009 02:54:38 -0000	1.1.1.1.4.2
+++ src/game/g_main.c	17 Feb 2009 03:11:02 -0000
@@ -182,6 +182,8 @@ vmCvar_t  g_mapvoteMaxTime;
 
 vmCvar_t  g_msg;
 vmCvar_t  g_msgTime;
+vmCvar_t  g_welcomeMsg;
+vmCvar_t  g_welcomeMsgTime;
 
 static cvarTable_t   gameCvarTable[ ] =
 {
@@ -344,6 +346,8 @@ static cvarTable_t   gameCvarTable[ ] =
   
   { &g_msg, "g_msg", "", CVAR_ARCHIVE, 0, qfalse  },
   { &g_msgTime, "g_msgTime", "0", CVAR_ARCHIVE, 0, qfalse  },
+  { &g_welcomeMsg, "g_welcomeMsg", "", CVAR_ARCHIVE, 0, qfalse  },
+  { &g_welcomeMsgTime, "g_welcomeMsgTime", "0", CVAR_ARCHIVE, 0, qfalse  },
   
   { &g_rankings, "g_rankings", "0", 0, 0, qfalse },
   { &g_allowShare, "g_allowShare", "0", CVAR_ARCHIVE | CVAR_SERVERINFO, 0, qfalse},
@@ -2438,6 +2442,38 @@ CheckMsgTimer
 */
 void CheckMsgTimer( void )
 {
+  static int LastTime = 0;
+
+  if( level.time - LastTime < 1000 )
+    return;
+
+  LastTime = level.time;
+
+  if( g_welcomeMsgTime.integer && g_welcomeMsg.string[ 0 ] )
+  {
+    char buffer[ MAX_STRING_CHARS ];
+    int wt;
+    int i;
+
+    buffer[ 0 ] = '\0';
+    wt = g_welcomeMsgTime.integer * 1000;
+    for( i = 0; i < level.maxclients; i++ )
+    {
+      if( level.clients[ i ].pers.connected != CON_CONNECTED )
+        continue;
+
+      if( level.time - level.clients[ i ].pers.enterTime < wt )
+      {
+        if( buffer[ 0 ] == '\0' )
+        {
+          Q_strncpyz( buffer, g_welcomeMsg.string, sizeof( buffer ) );
+          G_ParseEscapedString( buffer );
+        }
+      trap_SendServerCommand( i, va( "cp \"%s\"", buffer ) );
+      }
+    }
+  }
+
   if( !g_msgTime.integer )
     return;
 
